File: client-leger/app/src/main/java/com/services/general-chat.service.kt
package com.services

import ChatMessage
import ChatRoomMessageData
import constants.ChatRoomType
import constants.ChatSocketsEmitEvents
import constants.ChatSocketsSubscribeEvents
import JoinedChatRoom
import MessageInfo
import SocketService
import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import interfaces.ChatRoomInfo
import interfaces.OldMessage
import interfaces.toJson
import interfaces.toObject
import okio.IOException
import org.json.JSONObject
import toJson
import toObject


object ChatSocketService: Service() {
    private val _isChatOpen = MutableLiveData<Boolean>()
    val isChatOpen: LiveData<Boolean> get() = _isChatOpen
    private var channels = MutableLiveData<ArrayList<String>>(ArrayList())
    var joinedChatRooms = MutableLiveData<ArrayList<JoinedChatRoom>>(ArrayList())
    var currentChatRoom: JoinedChatRoom? = null

    private val _newMessage = MutableLiveData<ChatRoomMessageData?>()
    val newMessage: MutableLiveData<ChatRoomMessageData?> get() = _newMessage

    private val _changeChannel = MutableLiveData<Boolean>()
    val changeChannel: LiveData<Boolean> get() = _changeChannel

    val channelsSubject: LiveData<ArrayList<String>> get() = channels

    val joinedChatRoomsSubject: LiveData<ArrayList<JoinedChatRoom>> get() = joinedChatRooms

    private val socketService: SocketService = SocketService

    var hasUnreadMessages: Boolean = false

    //Section tp override Service
   override fun onBind(intent: Intent?): IBinder? {
        return null
   }

    init {
        _newMessage.postValue(null)
        _isChatOpen.postValue(false)
        currentChatRoom = null
        Runtime.getRuntime().addShutdownHook(Thread {
            try {
                SocketService.disconnect()
            }
            catch (e: IOException)
            {
                Log.e("ChatSocketService",e.toString())
            }
        })
    }

    fun connect() {
        SocketService.connect()
        Log.d("ChatSocketService", "connect() has been called.")
        setUpSocketListeners()
    }

    fun resetService(){
        _newMessage.postValue(null)
        _isChatOpen.postValue(false)
        currentChatRoom = null
        channels = MutableLiveData<ArrayList<String>>()
        joinedChatRooms = MutableLiveData<ArrayList<JoinedChatRoom>>(ArrayList())
        SocketService.disconnect()
    }

    private fun getChannels() {
        SocketService.send(ChatSocketsEmitEvents.GetChannels, null)
    }

    fun toggleChat() {
        if (_isChatOpen.value != null) {
            _isChatOpen.value = !_isChatOpen.value!!
            currentChatRoom = null
            _isChatOpen.postValue(_isChatOpen.value)
            _newMessage.postValue(null)
        } else {
            _isChatOpen.postValue(true)
        }
    }

    fun sendMessage(chatMsg: ChatRoomMessageData) {
        val msg = toJson(chatMsg)
        SocketService.send(ChatSocketsEmitEvents.SendMessage, msg)
    }

    fun selectChatRoom(chatRoomName: String) {
        val chatRoom = joinedChatRooms.value?.find { chatRoom -> chatRoom.chatRoomName == chatRoomName }
        Log.d("chatRooms", joinedChatRooms.value.toString())
        Log.d("chatRoomName", chatRoomName)
        _changeChannel.postValue(true)
        currentChatRoom = chatRoom
        Log.d("selectChannel", currentChatRoom.toString())
        chatRoom?.messages?.forEach { msg ->
            if (!msg.read) {
                msg.read = true
            }
        }
    }

    fun getUnreadMessageCount(channel: JoinedChatRoom): Number {
        return channel.messages.count { msg -> !msg.read }
    }

    fun backToChannels() {
        currentChatRoom = null
        this.newMessage.postValue(null)
    }

    private fun addMessageToChannel(msg: ChatRoomMessageData) {
        val chatroom = joinedChatRooms.value?.find{ c -> c.chatRoomName == msg.chatRoomName}
        if (chatroom != null) {
            addMessage(chatroom, msg.data)
        }
    }

    fun addMessage(channel: JoinedChatRoom, msg: ChatMessage) {
        channel.messages.add(MessageInfo(msg, channel == currentChatRoom))
            joinedChatRooms.postValue(joinedChatRooms.value)
            channels.postValue(channels.value)
    }

    fun setUpChannels(channels: JSONObject) {
        val channelsArray = channels.getJSONArray("channels")
        val joinedRoomsArray = channels.getJSONArray("joinedRooms")
        for(i in 0 until channelsArray.length())
        {
            val channel = channelsArray.getString(i)
            this.channels.value?.add(channel)
        }
        for(i in 0 until joinedRoomsArray.length())
        {
            val room = joinedRoomsArray.getJSONObject(i)
            val chatRoomName = room.getString("chatRoomName")
            if (!this.joinedChatRooms.value!!.any{chatRoom -> chatRoom.chatRoomName == chatRoomName}) {
                val messages = room.getJSONArray("messages")
                val chatRoomType: ChatRoomType = when (room.getString("chatRoomType")) {
                    "general" -> ChatRoomType.General
                    "public" -> ChatRoomType.Public
                    "match" -> ChatRoomType.Match
                    else -> throw(Error("invalid constants.ChatRoomType")) // or throw an exception if the type is not valid
                }
                val owner = room.getString("owner")
                val playersJson = room.getJSONArray("players")

                val messageInfos = ArrayList<MessageInfo>()
                for (j in 0 until messages.length()) {
                    val messageInfo = toObject(messages.getJSONObject(j), ChatMessage::class)
                    messageInfos.add(
                        MessageInfo(messageInfo, currentChatRoom?.chatRoomName == chatRoomName)
                    )
                }
                val players = ArrayList<String>()
                for (j in 0 until playersJson.length()){
                    players.add(playersJson.getString(j))
                }
                val joinedChatRoom =
                    JoinedChatRoom(chatRoomName, chatRoomType, owner, messageInfos, players)
                joinedChatRooms.value?.add(joinedChatRoom)
            }
        }
        this.channels.postValue(this.channels.value)
        joinedChatRooms.postValue(joinedChatRooms.value)
    }

    fun removeChannel(channelName: String) {
        joinedChatRooms.value = joinedChatRooms.value?.filter { joinedChatRoom -> joinedChatRoom.chatRoomName != channelName }
            ?.let { ArrayList(it) }
        channels.value = channels.value?.filter { channel -> channel != channelName }
            ?.let { ArrayList(it) }

        SocketService.send(ChatSocketsEmitEvents.LeaveChatRoom, channelName)
        joinedChatRooms.postValue(joinedChatRooms.value)
        channels.postValue(channels.value)
    }

    fun addChannel(channelName: String) {
        channels.value?.add(channelName)
        channels.postValue(channels.value)
    }

    fun joinChannel(chatRoomName: String) {
        SocketService.send(ChatSocketsEmitEvents.JoinChatRoom, chatRoomName)
    }

    fun createChannel(chatRoomInfo: ChatRoomInfo) {
        SocketService.send(ChatSocketsEmitEvents.CreateChatRoom, toJson(chatRoomInfo))
    }

    fun getMessages(): ArrayList<MessageInfo>? {
        return currentChatRoom?.messages
    }

    private fun setUpSocketListeners() {
        SocketService.on<String>(ChatSocketsSubscribeEvents.NewChatRoom) { name ->
            try {
                addChannel(name)
            } catch (error: Exception) {
                Log.e("ChatSocketService0", error.message ?: "Unknown Error")
            }
            Log.d("setUpSocketListeners", "Lambda2")
        }

        SocketService.on<String>(ChatSocketsSubscribeEvents.ChatClosed) { name ->
            removeChannel(name)
            Log.d("setUpSocketListeners", "Lambda3")
        }

        SocketService.on<JSONObject>(ChatSocketsSubscribeEvents.ChatJoined) { msg ->
            try {
                val chatRoomName = msg.getString("chatRoomName")
                if (joinedChatRooms.value!!.any{room -> room.chatRoomName == chatRoomName}) {
                    Log.d("ChatSocketService", "Already has $chatRoomName")
                } else {
                val messages = msg.getJSONArray("messages")
                Log.d("constants.ChatRoomType", msg.getString("chatRoomType"))
                Log.d("msg", msg.toString())
                val chatRoomType: ChatRoomType = when (msg.getString("chatRoomType")) {
                    "general" -> ChatRoomType.General
                    "public" -> ChatRoomType.Public
                    "match" -> ChatRoomType.Match
                    else -> throw (Error("invalid constants.ChatRoomType")) // or throw an exception if the type is not valid
                }
                val owner = msg.getString("owner")
                val playersJson = msg.getJSONArray("players")
                val messageInfos = ArrayList<MessageInfo>()
                for (i in 0 until messages.length()) {
                    val messageInfo = toObject(messages.getJSONObject(i), ChatMessage::class)
                    messageInfos.add(
                        MessageInfo(
                            messageInfo,
                            currentChatRoom?.chatRoomName == chatRoomName
                        )
                    )
                }
                val players = ArrayList<String>()
                for (i in 0 until playersJson.length()) {
                    players.add(playersJson.getString(i))
                }
                val joinedChatRoom =
                    JoinedChatRoom(chatRoomName, chatRoomType, owner, messageInfos, players)
                Log.d("ChatSocketService", "joinedChatRoom is '$joinedChatRoom'")
                joinedChatRooms.value?.add(joinedChatRoom)
                joinedChatRooms.postValue(joinedChatRooms.value)
                Log.d(
                    "ChatSocketService",
                    "joinedChatRooms values are '" + joinedChatRooms.value.toString() + "'"
                )
                }
            } catch (error: Exception) {
                Log.e("ChatSocketService1", error.message ?: "Unknown Error")
            }
        }

        SocketService.on<String?>(ChatSocketsSubscribeEvents.ChatLeft) { chatRoomName ->
            try {
                if (chatRoomName != null) {
                    channels.value?.add(chatRoomName)
                    channels.postValue(channels.value)
                }
            } catch (error: Exception) {
                Log.e("ChatSocketService2", error.message ?: "Unknown Error")
            }
        }

        SocketService.on<JSONObject>(ChatSocketsSubscribeEvents.ChatMessage) { msg ->
            try {
                val chatMessage = toObject(msg, ChatRoomMessageData::class)
                addMessageToChannel(chatMessage)
                hasUnreadMessages =
                    currentChatRoom?.chatRoomName !== chatMessage.chatRoomName || hasUnreadMessages
                Log.d(
                    "ChatSocketService",
                    currentChatRoom?.chatRoomName + " : " + chatMessage.chatRoomName
                )
                _newMessage.postValue(chatMessage)
               } catch (error: Exception) {
                Log.e("ChatSocketService", error.message ?: "Unknown Error")
            }
            Log.d("setUpSocketListeners", "Lambda1")
        }

        SocketService.on<JSONObject>(ChatSocketsSubscribeEvents.SendOldMessages) { msg ->
            try {
                val oldMessages = toObject(msg, OldMessage::class)
                val chatRoom =
                    joinedChatRooms.value?.find { joinedChatRoom -> joinedChatRoom.chatRoomName === oldMessages.chatRoomName }
                chatRoom?.let {
                    for (i in oldMessages.messages.indices.reversed()) {
                        it.messages.add(0, MessageInfo(oldMessages.messages[i], false))
                    }
                }
            } catch (error: Exception) {
                Log.e("ChatSocketService3", error.message ?: "Unknown Error")
            }
        }

        SocketService.on<JSONObject>(ChatSocketsSubscribeEvents.ChatRoomList) { cs ->
            try {
                setUpChannels(cs)
            } catch (error: Exception) {
                Log.e("ChatSocketService4", error.message ?: "Unknown Error")
            }
        }
    }
}

File: client/src/app/core/services/general-chat-service/general-chat.service.ts
import { EventEmitter, Injectable } from '@angular/core';
import { ChatSocketsEmitEvents, ChatSocketsSubscribeEvents } from '@app/core/constants/constants';
import { AccountService } from '@app/core/http/services/account-service/account.service';
import { ChatRoomMessage } from '@app/core/interfaces/chat-interfaces/chatroom-message';
import { ChatRoomMessageData } from '@app/core/interfaces/chat-interfaces/chatroom-message-data';
import { JoinedChatroom } from '@app/core/interfaces/chat-interfaces/joined-chatroom';
import { MappedMessages } from '@app/core/interfaces/chat-interfaces/mapped-messages';
import { OldMessages } from '@app/core/interfaces/chat-interfaces/old-messages';
import { ChatRoomInfo } from '@app/core/interfaces/chatroom-info';
import { SocketService } from '@app/core/websocket/services/socket-service/socket.service';
import { BehaviorSubject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class GeneralChatService {
    scrollToBottomEvent: EventEmitter<void> = new EventEmitter<void>();
    isChatOpen = false;
    currentChannel: JoinedChatroom | null = null;
    channels: string[] = [];
    joinedChannels: JoinedChatroom[] = [];
    isTyping: boolean = false;
    hasJustSentMessage: boolean = false;
    isChatClosed: boolean = false;
    mappedMessages: MappedMessages[] = [];
    // TODO: créer un service pour ces variables !!!!
    isProfileOpen: boolean = false;
    isFriendsOpen: boolean = false;
    isShopOpen: boolean = false;
    isSettingsOpen: boolean = false;
    private channelsSubject = new BehaviorSubject<string[]>(this.channels);
    private joinedChannelsSubject = new BehaviorSubject<JoinedChatroom[]>(this.joinedChannels);

    constructor(
        public socketService: SocketService,
        private accountService: AccountService,
    ) {
        socketService.connect();
        this.setupListeners(); // Set up socket listeners in the constructor
    }

    get channels$() {
        return this.channelsSubject.asObservable();
    }

    get joinedChannels$() {
        return this.joinedChannelsSubject.asObservable();
    }
    getChannels() {
        this.socketService.send(ChatSocketsEmitEvents.GetChannels);
    }

    toggleChat() {
        this.isChatOpen = !this.isChatOpen;
        if (!this.isChatOpen) {
            this.currentChannel = null;
        }
    }

    sendMessage(msg: ChatRoomMessageData) {
        this.socketService.send(ChatSocketsEmitEvents.SendMessage, msg);
    }

    async selectChannel(channel: JoinedChatroom): Promise<void> {
        this.currentChannel = channel;
        this.mappedMessages = [];
        if (channel.messages.length === 0) {
            return;
        }

        const userIds = channel.messages.map((msg) => msg.message.userId);

        try {
            const accounts = await this.accountService.getSomeAccounts(userIds).toPromise();

            const accountMap = new Map(accounts?.map((account) => [account.userId, account]));

            channel.messages.forEach((msg) => {
                if (!msg.read) {
                    msg.read = true;
                }

                const account = accountMap.get(msg.message.userId);
                if (account) {
                    this.mappedMessages.push({
                        playerName: account.pseudonym,
                        data: msg.message.data,
                        time: msg.message.time,
                        avatarUrl: this.accountService.getLocalAvatar(account.avatarUrl),
                        userId: msg.message.userId,
                        read: msg.read,
                    });
                }
            });

            this.mappedMessages.sort((a, b) => new Date(a.time).getTime() - new Date(b.time).getTime());
            this.scrollToBottomEvent.emit();
        } catch (error) {
            console.error('Error fetching accounts:', error);
        }
    }

    getUnreadMessageCount(chatroom: JoinedChatroom): number {
        return chatroom ? chatroom.messages.filter((msg) => !msg.read).length : 0;
    }

    backToChannels() {
        this.currentChannel = null;
    }

    addMessageToChannel(msg: ChatRoomMessageData) {
        const channel = this.joinedChannels.find((c) => c.chatRoomName === msg.chatRoomName);
        if (channel) {
            this.addMessage(channel, msg.data);
        }
    }

    addMessage(channel: JoinedChatroom, msg: ChatRoomMessage) {
        channel.messages.push({ message: msg, read: this.currentChannel === channel });
        if (this.currentChannel === channel) {
            this.accountService.getAccount(msg.userId).subscribe((account) => {
                this.mappedMessages.push({
                    playerName: account.pseudonym,
                    data: msg.data,
                    time: msg.time,
                    avatarUrl: this.accountService.getLocalAvatar(account.avatarUrl),
                    userId: msg.userId,
                    read: true,
                });
                this.scrollToBottomEvent.emit();
            });
        }

        this.channelsSubject.next(this.channels); // Notify subscribers
    }

    setUpChannels(channels: { joinedRooms: any[]; channels: string[] }) {
        console.log(channels);
        this.channels = channels.channels;
        this.joinedChannels = [];
        for (const channel of channels.joinedRooms) {
            const joinedChannel: JoinedChatroom = {
                chatRoomName: channel.chatRoomName,
                messages: channel.messages.map((message: ChatRoomMessage) => ({ message, read: false })),
                chatRoomType: channel.chatRoomType,
                owner: channel.owner,
                players: channel.players,
            };
            this.joinedChannels.push(joinedChannel);
        }
        this.channelsSubject.next(this.channels); // Notify
        this.joinedChannelsSubject.next(this.joinedChannels);
    }

    removeChannel(channelName: string) {
        this.joinedChannels = this.joinedChannels.filter((c) => c.chatRoomName !== channelName);
        this.channels = this.channels.filter((c) => c !== channelName);
        this.socketService.send(ChatSocketsEmitEvents.LeaveChatRoom, channelName);
        this.channelsSubject.next(this.channels);
        this.joinedChannelsSubject.next(this.joinedChannels);
    }

    addChannel(channel: string) {
        this.channels.push(channel);
        this.channelsSubject.next(this.channels); // Notify subscribers
    }

    joinChannel(chatRoomName: string) {
        console.log('joinChannel Called with: ' + chatRoomName);
        this.socketService.send(ChatSocketsEmitEvents.JoinChatRoom, chatRoomName);
        this.channelsSubject.next(this.channels); // Notify
        this.joinedChannelsSubject.next(this.joinedChannels);
    }

    createChannel(chatRoomInfo: ChatRoomInfo) {
        console.log('Creating channel: ' + chatRoomInfo.chatRoomName);
        this.socketService.send(ChatSocketsEmitEvents.CreateChatRoom, chatRoomInfo);
    }

    removeListeners() {
        this.socketService.removeListener(ChatSocketsSubscribeEvents.NewChatRoom);
        this.socketService.removeListener(ChatSocketsSubscribeEvents.ChatClosed);
        this.socketService.removeListener(ChatSocketsSubscribeEvents.ChatJoined);
        this.socketService.removeListener(ChatSocketsSubscribeEvents.ChatLeft);
        this.socketService.removeListener(ChatSocketsSubscribeEvents.ChatMessage);
        this.socketService.removeListener(ChatSocketsSubscribeEvents.SendOldMessages);
        this.socketService.removeListener(ChatSocketsSubscribeEvents.ChatRoomList);
    }

    private setupListeners() {
        this.socketService.on<string>(ChatSocketsSubscribeEvents.NewChatRoom, (chatRoomName: string) => {
            this.addChannel(chatRoomName);
        });

        this.socketService.on<string>(ChatSocketsSubscribeEvents.ChatClosed, (chatRoomName: string) => {
            this.removeChannel(chatRoomName);
        });

        this.socketService.on<any>(ChatSocketsSubscribeEvents.ChatJoined, (chatroom) => {
            console.log('joinedChannels : ');
            console.log(this.joinedChannels);
            console.log(chatroom);
            const joinedChannel: JoinedChatroom = {
                chatRoomName: chatroom.chatRoomName,
                messages: chatroom.messages.map((message: ChatRoomMessage) => ({ message, read: false })),
                chatRoomType: chatroom.chatRoomType,
                owner: chatroom.owner,
                players: chatroom.players,
            };
            this.joinedChannels.push(joinedChannel);
            this.channelsSubject.next(this.channels);
            this.joinedChannelsSubject.next(this.joinedChannels);
            console.log('The joined channels are : ');
            console.log(this.joinedChannels);
        });

        this.socketService.on<string>(ChatSocketsSubscribeEvents.ChatLeft, (chatRoomName: string | null) => {
            if (chatRoomName !== null) {
                this.channels.push(chatRoomName);
                this.channelsSubject.next(this.channels);
            }
        });

        this.socketService.on<ChatRoomMessageData>(ChatSocketsSubscribeEvents.ChatMessage, (msg: ChatRoomMessageData) => {
            this.addMessageToChannel(msg);
            this.hasJustSentMessage = true;
        });

        this.socketService.on<OldMessages>(ChatSocketsSubscribeEvents.SendOldMessages, (msg: OldMessages) => {
            const chatRoom = this.joinedChannels.find((channel) => channel.chatRoomName === msg.chatRoomName);
            if (chatRoom) {
                const messageInfo = msg.messages.map((m) => ({ message: m, read: false }));
                chatRoom.messages.unshift(...messageInfo);
            }
        });

        this.socketService.on<{ joinedRooms: any[]; channels: string[] }>(ChatSocketsSubscribeEvents.ChatRoomList, (msg) => {
            console.log('ChatRoomList : ');
            console.log(msg);
            this.setUpChannels(msg);
            console.log(this.channels);
        });
    }
}
